File: backend/src/test/java/com/skewwhiffy/auraltester/notation/factory/KeyFactoryTest.java
package com.skewwhiffy.auraltester.notation.factory;

import com.skewwhiffy.auraltester.test.util.TestData;

import static org.assertj.core.api.Assertions.assertThat;

import lombok.val;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import static org.mockito.Mockito.when;

import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class KeyFactoryTest {
    @Mock
    private NoteFactory noteFactory;
    @InjectMocks
    private KeyFactory keyFactory;

    @Test
    void instantiatesMajorKey() {
        val rawKey = "a#";
        val expectedNote = TestData.random().note();
        when(noteFactory.getNote("A#")).thenReturn(expectedNote);

        val actual = keyFactory.getKey(rawKey);

        assertThat(actual.note()).isEqualTo(expectedNote);
        assertThat(actual.isMinor()).isFalse();
    }

    @Test
    void recognizesMSuffixForMinor() {
        val rawKey = "abm";
        val expectedNote = TestData.random().note();
        when(noteFactory.getNote("Ab")).thenReturn(expectedNote);

        val actual = keyFactory.getKey(rawKey);

        assertThat(actual.note()).isEqualTo(expectedNote);
        assertThat(actual.isMinor()).isTrue();
    }

    @Test
    void recognizesMinorSuffixForMinor() {
        val rawKey = "bx minor";
        val expectedNote = TestData.random().note();
        when(noteFactory.getNote("Bx")).thenReturn(expectedNote);

        val actual = keyFactory.getKey(rawKey);

        assertThat(actual.note()).isEqualTo(expectedNote);
        assertThat(actual.isMinor()).isTrue();
    }
}

File: backend/src/test/kotlin/com/skewwhiffy/auraltester/notation/factory/KeyFactoryTest.kt
package com.skewwhiffy.auraltester.notation.factory

import com.ninjasquad.springmockk.MockkBean
import com.skewwhiffy.auraltester.test.util.TestData
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class KeyFactoryTest {
    @MockkBean
    private lateinit var noteFactory: NoteFactory

    @Autowired
    private lateinit var keyFactory: KeyFactory

    @Test
    fun instantiatesMajorKey() {
        val rawKey = "a#"
        val expectedNote = TestData.random.note
        every { noteFactory.getNote("A#") } returns expectedNote
        val actual = keyFactory.getKey(rawKey)
        assertThat(actual.note).isEqualTo(expectedNote)
        assertThat(actual.isMinor).isFalse()
    }

    @Test
    fun recognizesMSuffixForMinor() {
        val rawKey = "abm"
        val expectedNote = TestData.random.note
        every { noteFactory.getNote("Ab") } returns expectedNote
        val actual = keyFactory.getKey(rawKey)
        assertThat(actual.note).isEqualTo(expectedNote)
        assertThat(actual.isMinor).isTrue()
    }

    @Test
    fun recognizesMinorSuffixForMinor() {
        val rawKey = "bx minor"
        val expectedNote = TestData.random.note
        every { noteFactory.getNote("Bx") } returns expectedNote
        val actual = keyFactory.getKey(rawKey)
        assertThat(actual.note).isEqualTo(expectedNote)
        assertThat(actual.isMinor).isTrue()
    }
}
